mergify_engine/worker.py

```diff
@@ -209,38 +209,86 @@ def run(self):
+    _hash: str = dataclasses.field(init=False, default_factory=lambda: uuid.uuid4().hex)
+    _expire: int = 10
+
+    STREAM_GET_NEXT_AND_LOCK = """
+local now = ARGV[1]
+local count = ARGV[2]
+local hash = ARGV[3]
+local expire = ARGV[4]
+
+local streams = redis.call("ZRANGEBYSCORE", "streams", 0, now, "LIMIT", 0, count)
+for _, stream in ipairs(streams) do
+    local lock = "lock-" .. stream
+    local locked = redis.call("SET", lock, hash, "NX", "EX", expire)
+    if locked then
+        return stream
+    end
+end
+"""
+
+    STREAM_LOCK_UPDATE = """
+local lock = "lock-" .. KEYS[1]
+local hash = ARGV[1]
+local extend_time = ARGV[2]
+if redis.call("GET", lock) == hash then
+    return redis.call("SET", lock, hash, "EX", extend_time)
+else
+    return redis.error_reply('ERROR')
+end
+"""
+
+    STREAM_LOCK_RELEASE = """
+local lock = "lock-" .. KEYS[1]
+local hash = ARGV[1]
+if redis.call("GET", lock) == hash then
+    return redis.call("DEL", lock)
+else
+    return 0
+end
+"""
 
     @contextlib.asynccontextmanager
     async def next_stream(self):
-        # TODO(sileht): We can cache locally the result as the order is not going to
-        # change, and if it changes we don't care
-        # NOTE(sileht): We get the numbers we need to have one per worker, then remove
-        # streams already handled by other workers and keep the remaining one.
+        # NOTE(sileht): We get the numbers we need to have one per worker, they we try to
+        # acquire lock for each one until we found one free. We update the lock expiration
+        # in background, to ensure it will be release if the worker die
+        # TODO(sileht): don't HARDCODE self.process_count, do we really need it ? Can we
+        # known the number of dynos spawned ?
         now = time.time()
-        streams = [
-            s
-            for s in await self.redis.zrangebyscore(
-                "streams",
-                min=0,
-                max=now,
-                start=0,
-                num=self.worker_count * 2,
-            )
-            if s not in self._pending_streams
-        ]
-        if streams:
-            self._pending_streams.add(streams[0])
+        stream = await self.redis.eval(
+            self.STREAM_GET_NEXT_AND_LOCK,
+            0,
+            now,
+            self.local_worker_count * self.process_count * 2,
+            self._hash,
+            self._expire,
+        )
+        if stream:
+            lock_refresher = asyncio.create_task(self.refresh_lock(stream))
             try:
-                yield streams[0].decode()
+                yield stream.decode()
             finally:
-                self._pending_streams.remove(streams[0])
+                lock_refresher.cancel()
         else:
             yield
 
+    async def refresh_lock(self, stream):
+        while True:
+            try:
+                await self.redis.eval(
+                    self.STREAM_LOCK_UPDATE, 1, stream, self._hash, self._expire
```

Comment:
Looks like `STREAM_LOCK_UPDATE` return value is not checked. Maybe it should be so in case of collision at least we log or something?

Reply:
> What about using a lock free mechanism?
> You could select the stream based on a simple hash of the stream name modulo the worker number.
> That would spread the workload more or less evenly but would make sure it is lock free.

The spread will really depend on the used hash, the thing to hash is currently the org name.
Do you have any recommendation/lead for the hash method ?
I will read some stuff on this to find something in the meantime.

> The only question to answer is how to get the number of instances. That should be solvable.

Simple version is just the env with the expected workers. Other just a key continually refreshed in redis by each worker.